bitbucketserver: clean up plugin configuration

kzh opened the pull request:
This PR cleans up the Bitbucket Server plugin configurations. It moves all the feature options (alternative roaring bitmap permissions endpoint and automation webhooks) to the bitbucket server external service configuration on a per instance basis.

```
"plugin": {
  "webhooks": {
    "secret": "secret",
  },
  "permissions": "enabled"
}
```

The docs and changelog need to be updated.

unknwon commented on enterprise/cmd/frontend/internal/authz/bitbucketserver/provider.go:
```diff
@@ -26,6 +25,8 @@ type Provider struct {
 	pageSize int // Page size to use in paginated requests.
 	store    *store
+
+	fastPerm bool
```
Add a comment for this field would be nice.

unknwon commented on internal/conf/computed.go:
```diff
@@ -265,14 +265,6 @@ func SearchSymbolsParallelism() int {
 	return val
 }
 
-func BitbucketServerFastPerm() bool {
```
I thought having this helper function is actually nicer?

kzh replied:
This config was moved out of the site config and into the individual bitbucket server external service configs, so I inlined the logic when creating the authz provider.

kzh replied:
This was added back in

tsenart commented on schema/bitbucket_server.schema.json:
```diff
@@ -59,14 +59,34 @@
+              "type": "string",
+              "minLength": 1
+            },
+            "automation": {
```
What makes this setting necessary / valuable?

kzh replied:
I added this setting to give site admins the ability to toggle/disable webhook updates for specific features (ex: a8n). At the moment, there is only a8n but I assumed that webhooks could be used for something else down the line. This might not be ideal right now.

tsenart replied:
I don't think we need this toggle. It's good for us to control what we do with webhooks. The admin can just decide not to use webhooks at all, but not which product features leverage them.

kzh replied:
Gotcha, that makes sense!

tsenart commented on schema/bitbucket_server.schema.json:
```diff
@@ -59,14 +59,34 @@
     },
-    "webhooks": {
-      "description": "Configuration for Bitbucket Server Sourcegraph plugin webhooks",
+    "plugin": {
```
This config change is not backwards compatible. I think it'd break things for existing customers with webhooks enabled.

kzh replied:
Ah right. Is there an existing policy on how this should be handled? I was thinking of keeping the the old setting here and changing the description to "deprecated". However, there isn't much visibility here for site admins.

tsenart replied:
> Is there migrations for config schema changes?

Not that I'm aware. Can you make this change backwards compatible instead?

tsenart commented on schema/site.schema.json:
```diff
@@ -85,12 +85,6 @@
-          "type": "string",
-          "enum": ["enabled", "disabled"],
-          "default": "disabled"
-        },
```
This is also not backwards compatible.

keegancsmith commented on schema/bitbucket_server.schema.json:
```diff
@@ -59,14 +59,34 @@
+          }
+        },
+        "fastPerm": {
+          "description": "Enables fetching Bitbucket Server permissions through the roaring bitmap endpoint. Warning: there may be performance degradation under significant load.",
```
this warning should link to some more information.

kzh replied:
I'll address this in a followup PR.

tsenart replied:
> I'll address this in a followup PR.

Why not in this PR?

kzh replied:
There are any additional documents detailing the performance degradation at the moment. I plan to write extensive docs (sourcegraph/customer#18).

tsenart replied:
>  I plan to write extensive docs for sourcegraph/customer#18.

Oh! When do you plan to have these docs done?

tsenart replied:
On second thought, I think it'd be better to have a longer name that is more descriptive. `fastPerm` is too hacky a name for admins. How about just `permissions`? Since it is scoped to `plugin`, I think that'd be fine?

kzh commented:
I added back in the previously removed settings and changed their descriptions to indicate deprecation. Both settings are supported, but the newer one is recommended.

tsenart commented on schema/bitbucket_server.schema.json:
```diff
@@ -70,6 +70,31 @@
+        "fastPerm": {
+          "description": "Enables fetching Bitbucket Server permissions through the roaring bitmap endpoint. Warning: there may be performance degradation under significant load.",
+          "type": "string",
+          "enum": ["enabled", "disabled"],
```
Do we have prior art on using enabled and disabled enum instead of a boolean type?

kzh replied:
This carried over from the experimental features section in the site config.

unknwon replied:
Unfortunately, we seem having both styles: https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+file:%5Eschema/+%22enabled%22&patternType=literal

keegancsmith replied:
rather use boolean. enum exists from before our go library for schema supported a way to detect if the value is set. We now have a way with `"!go": { "pointer": true }`. Note: if the default is false/disabled then you don't need the pointer to detect unset. Also note: enabled/disabled enum is useful in the case you want to add new values (such as auto, etc). This shouldn't happen in this case, but is more common than you may think.

tsenart commented on internal/extsvc/bitbucketserver/client.go:
```diff
@@ -92,6 +95,18 @@ func NewClient(url *url.URL, httpClient httpcli.Doer) *Client {
 }
 
+// Secret returns the webhook secret from a BBS config
+func Secret(c *schema.BitbucketServerConnection) string {
```
Is this the right package for this? Can't we add this as a method to the `schema.BitbucketServerConnection` type? It'd be just adding a go file to that directory that would extend the type.

kzh replied:
Got it! I previously avoided this because the the BitbucketServerConnection type is generated from the JSON schema, but this is a much better placement.

tsenart commented on enterprise/cmd/frontend/internal/authz/bitbucketserver/provider.go:
```diff
@@ -26,6 +25,10 @@ type Provider struct {
 	store    *store
+
+	// fastPerm enables fetching permissions from the alternative roaring
+	// bitmap endpoint provided by the plugin
```
What plugin? Let's be more explicit and link to it.

tsenart commented on internal/extsvc/bitbucketserver/client.go:
```diff
@@ -65,6 +65,9 @@ type Client struct {
 	Username, Password string
 
+	// The secret used to create and authenticate plugin webhooks
+	WebhookSecret string
```
Is this a leftover?

keegancsmith commented on schema/bitbucket_server_utils.go:
```diff
@@ -0,0 +1,13 @@
+	default:
+		return ""
+	}
+}
```
nice

tsenart commented:
What's the status on this? Can it be merged?

kzh commented:
Made a quick change `s/fastPerm/pluginPerm` and merged!
